src/System.Private.CoreLib/shared/System/String.cs

```diff
@@ -22,7 +23,7 @@ namespace System
 
     [Serializable]
     [System.Runtime.CompilerServices.TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
-    public sealed partial class String : IComparable, IEnumerable, IConvertible, IEnumerable<char>, IComparable<string?>, IEquatable<string?>, ICloneable
+    public sealed partial class String : IComparable, IEnumerable, IConvertible, IEnumerable<char>, IComparable<string>, IEquatable<string>, ICloneable
```

Comment:
Did you remove this ? and instead use the attributes in the APIs? This to avoid having mismatch of types when doing typeof and stuff like that?

Reply:
I see that there is an issue for covariant interface parameter types.

Reply:
> Did you remove these ? and instead use the attributes in the APIs? This to avoid having mismatch of types when doing typeof and stuff like that?

It's because of a problem I ran into, and this was the best result @MadsTorgersen and I came up with.  If I have:
```C#
class Foo<T> where T : IEquatable<T>
```
then if `String` is defined to be an `IEquatable<string?>`, and if you try to use `Foo<string>`, then you effectively end up with:
```C#
class Foo<string> where string : IEquatable<string>
```
but `string` isn't `IEquatable<string>`, it's `IEquatable<string?>`, and the compiler complains:
```
error CS8631: The type 'string' cannot be used as type parameter 'T' in the generic type or method 'GenericEqualityComparer<T>'. Nullability of type argument 'string' doesn't match constraint type 'System.IEquatable<string>'
```
So basically, anytime we have a reference type that implements a generic interface with itself as the type parameter, we should aim for it to use a non-nullable version of itself (which logically makes sense).  The interface then uses attributes to express the nullability of each argument.

Not ideal, but the best we came up with.

Reply:
Makes total sense. Thanks for the detailed explanation.